src/irmin-pack/unix/ext.ml

```diff
@@ -233,29 +286,181 @@ module Maker (Config : Conf.S) = struct
+              let open Result_syntax in
+              (* Unlink previous suffix. *)
+              let suffix =
+                Irmin_pack.Layout.V3.suffix ~root ~generation:(generation - 1)
+              in
+              let* () = Io.unlink suffix in
+              (* Unlink current gc's result.*)
+              let result = Irmin_pack.Layout.V3.gc_result ~root ~generation in
+              Io.unlink result
+            in
+            match result with
+            | Error e ->
+                [%log.warn
+                  "Unlinking temporary files after gc failed with error %a"
+                    Errs.pp e]
+            | Ok () -> ()
+
+          let gc_errors (status, gc_output) =
+            match (status, gc_output) with
+            | Lwt_unix.WEXITED n, Error (`Gc_process_error str) ->
+                Error (`Gc_process_error (Fmt.str "Exited %d %s" n str))
+            | Lwt_unix.WEXITED n, Error (`Corrupted_gc_result_file str) ->
+                Error (`Corrupted_gc_result_file (Fmt.str "Exited %d %s" n str))
+            | Lwt_unix.WSIGNALED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Signaled %d" n))
+            | Lwt_unix.WSTOPPED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d" n))
+            | Lwt_unix.WSIGNALED n, Error (`Gc_process_error str) ->
+                Error (`Gc_process_error (Fmt.str "Signaled %d %s" n str))
+            | Lwt_unix.WSTOPPED n, Error (`Gc_process_error str) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d %s" n str))
+            | Lwt_unix.WEXITED _, Ok _
+            | Lwt_unix.WSIGNALED _, Ok _
+            | Lwt_unix.WSTOPPED _, Ok _ ->
+                assert false
+
+          let finalise ~wait t =
+            match t.during_gc with
+            | None -> Lwt.return_ok false
+            | Some { next_generation; pid; unlink } ->
+                let wait_flags = if wait then [] else [ Unix.WNOHANG ] in
+                let root = Conf.root t.config in
+                let go status =
+                  let gc_output =
+                    File_manager.read_gc_output ~root
+                      ~generation:next_generation
+                  in
+                  let result =
+                    let open Result_syntax in
+                    match (status, gc_output) with
+                    | Lwt_unix.WEXITED 0, Ok copy_end_offset ->
+                        let* () =
+                          copy_latest_newies ~generation:next_generation
+                            ~copy_end_offset ~root t
+                        in
+                        let* () =
+                          swap_and_purge ~generation:next_generation
+                            ~copy_end_offset t
```

Comment:
it shouldn't be `copy_end_offset` but the offset returned by `copy_latest_newies`